pkg/agent/openflow/pipeline.go

```diff
@@ -2163,6 +2163,7 @@ func (c *client) snatCommonFlows(nodeIP net.IP, localSubnet net.IPNet, localGate
 		L3ForwardingTable.BuildFlow(priorityLow).
 			MatchProtocol(ipProto).
 			MatchRegMark(FromLocalRegMark).
+			MatchCTMark(NotServiceCTMark). // Don't match Service packets from local Pods.
```

Comment:
This doesn't seem right. Service and Egress are not mutually exclusive:
- A packet accessing a service whose backend pod is an external address should enforce Egress SNAT as well.
- An inbound NodePort request handled by kube-proxy doesn't have ServiceCTMark set, but should not enforce Egress SNAT.

The real rule is: only request packets should enforce Egress, reply packets shouldn't.

Reply:
You are right, I have updated the commit and could you have a look at the updated code to see that if it is right. I'll add e2e test in latter commit. Thanks.

Reply:
I think it may be hard to test this in the e2e. It needed to test the NodePort connection outside the Kubernetes cluster.